### ISSUE TITLE
Want to make Improvements

### ISSUE THREAD

**Issue by mpkorstanje** [nickmancol/simmetrics#2]
Hey, I've been looking through the code some more and I would like to make the following improvements. I want to do this because I want to practice my design and refactoring skills.

If you make these changes, are you interested in incorporating them? 

# Part 1 #

**Support Caching**

When doing a `AbsractStringMetric.batchCompareSet()` the tokenizer will repeatedly have to tokenize the same string. By creating a caching tokenizer with a cache of size two the and eviction policy of least used this will increase performance. A caching tokenizer will act as a decorator for the regular tokenizer.

**Explicit String normalisation**

Some algorithms such as SimonWhite and Soundex are using implicit string normalisation. All upper case, no spaces, ect. This too can be handled by a tokenizer decorator. This allows different normalisation schemes to be used.

Also in combination with a caching tokenizer this can then be used to increase performance again because the same string doesn't have to be normalized over and over again. 

**Clean up comments**

The current comments are auto-generated and often don't provide information. I would like to remove all identical comments from the algorithms just document the interfaces and abstract classes. Most IDE's will pick the documentation from the interface/abstract class when the concrete class has none available on an overriden method.

**Introduce a standard unit tests for all methods on InterfaceStringMetric**

Right now the same unit test is duplicated for each metric and only tests one method. This isn't really needed.

# Part 2 #

This part is a bit more rigorous and will include API changes.

**Performance measuring decorator**

Measuring the performance of a string comparison isn't really part of the core functionality of a string metric. This can be moved to a decorator. 

**Remove non functional methods**

Some method such as `getLongDescription` and `getSimilarityExplained` aren't implemented by most metrics, the method just returns null. The `getShortDescription` often just returns the simple class name. 

Removing these removes clutter from the library. 

**Split library into metrics and performance testing modules**

I would like to split the project into two sub modules. One module that has just the the metrics and an other which does the performance testing. This allows people to include just metrics in their project without having to look through the performance testing classes.

**Rename InterfaceStringMetric to StringMetric**

Java standard names don't include interface in the name. When it happens it usually means that your interface doesn't add anything to a concrete class. Since this clearly not the case it can be renamed.

**Comment by mpkorstanje** [comment]
Okay. I reckon I've got something reasonable stable now.

It took me a while to figure out how to separate all the concerns. Had to realize that some metrics operate on strings, some on lists of strings and some on sets of strings and that this makes them different metrics.

To turn these back into string metrics a tokenizer is mandatory. The old implementation did this by slapping a whitespace tokenizer on nearly everything but that doesn't always make sense for every application.

This in itself wasn't a problem. It was easy to replace the whitespace tokenzier with something else. But doing so in combination with other metrics that didn't use a white space tokenizer and actually depend on a specific tokenization scheme was playing havoc with the caching support I had in mind.

Its on the version2 branch in my repository. I've increased the numbers in the POM to match. It is just not possible to do this and keep it backwards compatible.
